packages/contracts/src/core/permission/PermissionManager.sol

```diff
@@ -109,6 +182,213 @@ abstract contract PermissionManager is Initializable {
+
+    /// @dev Requires the `ROOT_PERMISSION_ID` permission.
+    /// @param _where The address of the target contract for which `_who` receives permission.
+    /// @param _permissionIdOrSelector The permission hash or function selector used for this permission.
+    /// @param _owner The initial owner of this newly created permission.  
+    /// @param _whos The addresses of the target contracts for which `_who` receives permission.
+    function createPermission(
+        address _where,
+        bytes32 _permissionIdOrSelector,
+        address _owner,
+        address[] calldata _whos
+    ) external auth(ROOT_PERMISSION_ID) {
+        _createPermission(_where, _permissionIdOrSelector, _owner, _whos);
+    }
+
+    /// @notice Function to delegate specific flags of a permission. 
+    /// @param _where The address of the target contract for which `_who` receives permission.
+    /// @param _permissionIdOrSelector The permission hash or function selector used for this permission.
+    /// @param _delegatee The addresses who gets the permission delegated.
+    /// @param _flags The flags as uint8 the permission owner wants to give this specific delegatee.
+    function delegatePermission(
+        address _where,
+        bytes32 _permissionIdOrSelector,
+        address _delegatee,
+        uint8 _flags
+    ) public {
+        bytes32 permHash = permissionHash(_where, _permissionIdOrSelector);
+        Permission storage permission = permissions[permHash];
+
+        if (_isPermissionFrozen(permission)) {
+            revert PermissionFrozen(_where, _permissionIdOrSelector);
+        }
+
+        if (!hasPermission(permission.owners[msg.sender], _flags)) {
+            revert InvalidOwnerPermission(msg.sender, permission.owners[msg.sender], _flags);
+        }
+
+        uint8 newFlags = permission.delegations[_delegatee][permHash] | _flags;
+        permission.delegations[_delegatee][permHash] = newFlags;
+
+        emit PermissionDelegated(_where, _permissionIdOrSelector, _delegatee, newFlags);
+    }
+
+    /// @notice Function to remove sepcific flags from the delegatee
+    /// @param _where The address of the target contract for which `_who` receives permission.
+    /// @param _permissionIdOrSelector The permission hash or function selector used for this permission.
+    /// @param _delegatee The addresses we want to undelegate specifc flags.
+    /// @param _flags The flags as uint8 the permission owner wants to remove from this specific delegatee.
+    function undelegatePermission(
+        address _where,
+        bytes32 _permissionIdOrSelector,
+        address _delegatee,
+        uint8 _flags
+    ) public {
+        bytes32 permHash = permissionHash(_where, _permissionIdOrSelector);
+        Permission storage permission = permissions[permHash];
+
+        if (!hasPermission(permission.owners[msg.sender], _flags)) {
+            revert InvalidOwnerPermission(msg.sender, permission.owners[msg.sender], _flags);
+        }
+
```

Comment:
There needs to be a check if the permission is really delegated to `_delegatee` or not. If not, we shouldn't emit the event and revert. 

We might decide to silently succeed(i.e not revert), but this also might have consequences.